File: server/priorityEngine.ts
/**
 * PRISM Priority Engine (Phase 3A)
 *
 * Deterministically turns project risk assessments into an evidence-based
 * Intervention Priority Queue for infrastructure decision-makers.
 *
 * Core Principle:
 *   Priority is NOT simply riskScore.
 *   High risk alone does not imply operational urgency (e.g. a high-risk project
 *   with a completion date 5 years away does not require immediate cabinet intervention).
 *   The intervention queue prioritizes projects where high risk is BOTH credible
 *   (confirmed by longitudinal evidence) AND operationally urgent (deadline pressure,
 *   active stagnation, or acute deterioration).
 *
 * Deterministic Priority Formula:
 *   Priority Score = round(
 *       0.40 * RiskScore
 *     + 0.25 * ScheduleUrgency
 *     + 0.20 * RecentProgressDeterioration
 *     + 0.15 * (EvidenceConfidence * 100)
 *   )
 *   Bounded in [0, 100].
 *
 * Weights (Sum = 100%):
 *   1. PRISM Risk Index                 : 40% (0.40)
 *   2. Schedule Urgency                 : 25% (0.25)
 *   3. Recent Progress Deterioration    : 20% (0.20)
 *   4. Evidence Confidence              : 15% (0.15)
 *
 * Priority Tiers:
 *   P1 (Immediate intervention)        : Priority Score >= 70
 *   P2 (High-priority monitoring)       : 50 <= PriorityScore < 70
 *   P3 (Routine monitoring)             : Priority Score < 50
 *
 * Recommended Actions:
 *   Deterministic, indicator-grounded rules. No ML model claims.
 */

import { PaimanaObservation, RiskTier } from '../src/types/index';
import { PRISMRiskEngine, RiskAssessment } from './riskEngine';

export type PriorityTier = 'P1' | 'P2' | 'P3';

export interface PriorityAssessment {
  projectId: string;
  projectName: string;
  code: string;
  sector: string;
  state: string;
  implementingAgency: string;
  priorityScore: number;
  priorityTier: PriorityTier;
  riskScore: number;
  riskTier: RiskTier;
  evidenceConfidence: number; // 0.25 to 1.00
  urgency: number; // 0 to 100
  recentDeterioration: number; // 0 to 100
  primaryRiskDriver: string;
  topRiskDrivers: string[];
  priorityReason: string;
  recommendedAction: string;
  evidence: {
    physicalProgressPercent: number;
    cumulativeExpenditureCr: number;
    revisedCostCr: number;
    originalCostCr: number;
    costOverrunPercent: number;
    expenditurePctOfRevisedCost: number;
    timeOverrunMonths: number;
    targetCompletionDate: string;
    originalCompletionDate: string;
    consecutiveStagnantMonths: number;
    recentProgressDelta: number | null;
    observationCount: number;
  };
  assessedAt: string;
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}

function parseCompletionDateToFractionalYear(str?: string): number | null {
  if (!str) return null;
  const parts = str.trim().split('/');
  if (parts.length !== 2) return null;
  const m = parseInt(parts[0], 10);
  const y = parseInt(parts[1], 10);
  if (isNaN(m) || isNaN(y)) return null;
  return y + (m - 1) / 12;
}

function reportMonthToFractionalYear(s: string): number {
  const parts = s.split('-');
  if (parts.length !== 2) return 0;
  const y = parseInt(parts[0], 10);
  const m = parseInt(parts[1], 10);
  return y + (m - 1) / 12;
}

/**
 * Calculates Schedule Urgency (0-100).
 *
 * Operational urgency measures deadline proximity and overdue status:
 *  - Past deadline (remainingYears <= 0): 85-100 (inversely proportional to progress)
 *  - Imminent (0 < remainingYears <= 0.5, within 6 months): 70-95
 *  - Approaching (0.5 < remainingYears <= 1.0, 6-12 months): 50-70
 *  - Medium term (1.0 < remainingYears <= 2.0, 1-2 years): 30-50
 *  - Long horizon (remainingYears > 2.0): 5-30
 *  - If project is physically 100% complete: 0
 *  - If no completion date recorded: 40 (moderate default)
 */
export function calcScheduleUrgency(latest: PaimanaObservation): number {
  const physPct = latest.physical_progress_pct ?? 0;
  if (physPct >= 100) return 0;

  const targetDateStr = latest.revised_target_completion_mm_yyyy || latest.original_target_completion_mm_yyyy;
  const targetYear = parseCompletionDateToFractionalYear(targetDateStr);
  if (targetYear == null) {
    return 40; // Neutral baseline if no target completion date is provided
  }

  const reportYear = reportMonthToFractionalYear(latest.report_month || '2026-07');
  const remainingYears = targetYear - reportYear;

  let urgency: number;
  if (remainingYears <= 0) {
    // Already overdue
    urgency = 85 + (100 - physPct) * 0.15;
  } else if (remainingYears <= 0.5) {
    // Within 6 months
    const timeFactor = (1 - remainingYears / 0.5) * 15;
    const progressFactor = ((100 - physPct) / 100) * 10;
    urgency = 70 + timeFactor + progressFactor;
  } else if (remainingYears <= 1.0) {
    // 6 to 12 months
    urgency = 50 + ((1.0 - remainingYears) / 0.5) * 20;
  } else if (remainingYears <= 2.0) {
    // 1 to 2 years
    urgency = 30 + ((2.0 - remainingYears) / 1.0) * 20;
  } else {
    // > 2 years
    urgency = 30 - (remainingYears - 2.0) * 5;
  }

  return clamp(Math.round(urgency), 0, 100);
}

/**
 * Calculates Recent Progress Deterioration (0-100) and consecutive stagnant months.
 *
 * Evaluates whether project progress is stalling, retrogressing, or sharply decelerating.
 *  - Retrogression (delta < 0): 100
 *  - Stagnant (delta == 0):
 *      >= 3 consecutive stagnant intervals: 95
 *      2 consecutive stagnant intervals: 80
 *      1 stagnant interval: 65
 *  - Sluggish (0 < delta < 0.5 pp/mo): 40-55
 *  - Moderate (0.5 <= delta < 1.0 pp/mo): 25
 *  - Adequate (1.0 <= delta < 2.0 pp/mo): 10
 *  - Healthy (delta >= 2.0 pp/mo): 0
 *  - Sudden deceleration bonus: +15 if historical early velocity was high but recent velocity dropped
 */
export function calcRecentDeterioration(obs: PaimanaObservation[]): { score: number; consecutiveStagnantMonths: number; recentDelta: number | null } {
  if (!obs || obs.length < 2) {
    return { score: 30, consecutiveStagnantMonths: 0, recentDelta: null };
  }

  const deltas: number[] = [];
  for (let i = 1; i < obs.length; i++) {
    deltas.push((obs[i].physical_progress_pct ?? 0) - (obs[i - 1].physical_progress_pct ?? 0));
  }

  const recentDelta = deltas[deltas.length - 1];

  // Count consecutive stagnant intervals from the end
  let consecutiveStagnantMonths = 0;
  for (let i = deltas.length - 1; i >= 0; i--) {
    if (deltas[i] <= 0.1) {
      consecutiveStagnantMonths++;
    } else {
      break;
    }
  }

  let score: number;
  if (recentDelta < 0) {
    score = 100;
  } else if (recentDelta <= 0.1) {
    if (consecutiveStagnantMonths >= 3) score = 95;
    else if (consecutiveStagnantMonths === 2) score = 80;
    else score = 65;
  } else if (recentDelta < 0.5) {
    score = consecutiveStagnantMonths > 0 ? 55 : 40;
  } else if (recentDelta < 1.0) {
    score = 25;
  } else if (recentDelta < 2.0) {
    score = 10;
  } else {
    score = 0;
  }

  // Trend deceleration penalty
  if (obs.length >= 3) {
    const earlyDeltas = deltas.slice(0, Math.floor(deltas.length / 2));
    const earlyAvg = earlyDeltas.reduce((s, d) => s + d, 0) / (earlyDeltas.length || 1);
    const decel = earlyAvg - recentDelta;
    if (decel > 2.0) {
      score += 15;
    }
  }

  return {
    score: clamp(Math.round(score), 0, 100),
    consecutiveStagnantMonths,
    recentDelta: parseFloat(recentDelta.toFixed(2))
  };
}

/**
 * Assigns Priority Tier from Priority Score:
 *   P1 = Immediate intervention    (Score >= 70)
 *   P2 = High-priority monitoring   (50 <= Score < 70)
 *   P3 = Routine monitoring         (Score < 50)
 */
export function priorityTierFromScore(score: number): PriorityTier {
  if (score >= 70) return 'P1';
  if (score >= 50) return 'P2';
  return 'P3';
}

/**
 * Deterministically generates recommended actions based on actual indicator evidence.
 * No ML model claims.
 */
export function buildRecommendedAction(
  assessment: RiskAssessment,
  urgency: number,
  recentDeterioration: number,
  consecutiveStagnantMonths: number,
  costOverrunPercent: number,
  spendProgressGap: number
): string {
  const actions: string[] = [];

  // Indicator lookup
  const getIndicatorScore = (id: string) => assessment.indicators.find(i => i.id === id)?.normalisedScore ?? 0;

  const stagnationScore = getIndicatorScore('stagnation');
  const costScore = getIndicatorScore('cost');
  const scheduleScore = getIndicatorScore('schedule');
  const divergenceScore = getIndicatorScore('divergence');
  const velocityScore = getIndicatorScore('velocity');

  // Rule 1: Prolonged Stagnation
  if (consecutiveStagnantMonths >= 2 || stagnationScore >= 60) {
    actions.push('Escalate physical-progress review');
  }

  // Rule 2: High Schedule Pressure
  if (urgency >= 75 || scheduleScore >= 60) {
    actions.push('Review completion recovery plan');
  }

  // Rule 3: Expenditure substantially ahead of physical progress (Divergence)
  if (divergenceScore >= 60 || spendProgressGap >= 15) {
    actions.push('Review expenditure–physical progress divergence');
  }

  // Rule 4: Severe Cost Escalation
  if (costScore >= 60 || costOverrunPercent >= 20) {
    actions.push('Initiate cost/revision review');
  }

  // Rule 5: Severe Velocity Deceleration / Contractor Sluggishness
  if (recentDeterioration >= 70 || velocityScore >= 70) {
    if (!actions.includes('Escalate physical-progress review')) {
      actions.push('Conduct contractor mobilization & site execution audit');
    }
  }

  if (actions.length === 0) {
    return 'Maintain routine milestone & expenditure monitoring';
  }

  return actions.join('; ');
}

/**
 * Builds deterministic human-readable reason for priority assignment.
 */
export function buildPriorityReason(
  tier: PriorityTier,
  priorityScore: number,
  riskScore: number,
  urgency: number,
  deterioration: number,
  confidence: number,
  consecutiveStagnantMonths: number,
  topDrivers: string[]
): string {
  const confidencePercent = Math.round(confidence * 100);

  if (tier === 'P1') {
    const reasons: string[] = [];
    if (urgency >= 75) reasons.push(`high schedule urgency (${urgency}/100)`);
    if (consecutiveStagnantMonths >= 2) reasons.push(`${consecutiveStagnantMonths} consecutive months of stagnant progress`);
    else if (deterioration >= 60) reasons.push(`acute recent progress deterioration (${deterioration}/100)`);
    if (reasons.length === 0 && topDrivers.length > 0) reasons.push(`critical ${topDrivers[0].toLowerCase()}`);

    const driverDetail = reasons.length > 0 ? reasons.join(' and ') : 'compounded operational risk indicators';
    return `P1 Immediate Intervention: High PRISM Risk Index (${riskScore}) confirmed with ${confidencePercent}% evidence confidence, compounded by ${driverDetail}.`;
  }

  if (tier === 'P2') {
    return `P2 High-Priority Monitoring: Elevated composite score (${priorityScore}) with operational urgency (${urgency}/100); key indicators (${topDrivers.slice(0, 2).join(', ') || 'execution'}) warrant focused oversight.`;
  }

  return `P3 Routine Monitoring: Project maintains manageable priority index (${priorityScore}) with risk index ${riskScore} and low immediate schedule pressure.`;
}

export class PRISMPriorityEngine {
  /**
   * Weights configuration:
   *   PRISM Risk Index:              40%
   *   Schedule Urgency:              25%
   *   Recent Progress Deterioration: 20%
   *   Evidence Confidence:           15%
   */
  public static readonly WEIGHTS = {
    risk: 0.40,
    urgency: 0.25,
    deterioration: 0.20,
    confidence: 0.15
  } as const;

  /**
   * Calculates the deterministic Priority Assessment for a project.
   *
   * @param observations  Sorted PAIMANA monthly snapshots for the project.
   * @param existingAssessment Optional precomputed RiskAssessment from PRISMRiskEngine.
   */
  public static assess(
    observations: PaimanaObservation[],
    existingAssessment?: RiskAssessment
  ): PriorityAssessment {
    if (!observations || observations.length === 0) {
      throw new Error('PRISMPriorityEngine.assess() requires at least one observation');
    }

    const obs = [...observations].sort((a, b) => a.report_month.localeCompare(b.report_month));
    const latest = obs[obs.length - 1];

    // 1. PRISM Risk Index & Assessment
    const riskAssessment = existingAssessment || PRISMRiskEngine.assess(obs);
    const riskScore = riskAssessment.riskScore;
    const riskTier = riskAssessment.riskTier;
    const evidenceConfidence = riskAssessment.evidenceConfidence; // 0.25 to 1.00

    // 2. Schedule Urgency (0 to 100)
    const urgency = calcScheduleUrgency(latest);

    // 3. Recent Progress Deterioration (0 to 100)
    const { score: recentDeterioration, consecutiveStagnantMonths, recentDelta } = calcRecentDeterioration(obs);

    // 4. Evidence Confidence normalized (0 to 100)
    const confidenceScore = clamp(evidenceConfidence * 100, 0, 100);

    // Deterministic Priority Formula:
    // PriorityScore = 0.40 * riskScore + 0.25 * urgency + 0.20 * deterioration + 0.15 * confidenceScore
    const rawPriorityScore = (
      riskScore * PRISMPriorityEngine.WEIGHTS.risk +
      urgency * PRISMPriorityEngine.WEIGHTS.urgency +
      recentDeterioration * PRISMPriorityEngine.WEIGHTS.deterioration +
      confidenceScore * PRISMPriorityEngine.WEIGHTS.confidence
    );

    const priorityScore = clamp(Math.round(rawPriorityScore), 0, 100);
    const priorityTier = priorityTierFromScore(priorityScore);

    // Drivers
    const sortedIndicators = [...riskAssessment.indicators].sort(
      (a, b) => b.weightedContribution - a.weightedContribution
    );
    const primaryRiskDriver = sortedIndicators[0]?.label || 'General Execution';
    const topRiskDrivers = sortedIndicators
      .filter(ind => ind.normalisedScore >= 25)
      .slice(0, 3)
      .map(ind => ind.label);

    if (topRiskDrivers.length === 0) {
      topRiskDrivers.push(primaryRiskDriver);
    }

    // Cost Overrun and Gap Calculation
    const originalCost = latest.original_cost_cr || 0;
    const revisedCost = latest.revised_cost_cr || originalCost;
    const costOverrunPercent = originalCost > 0
      ? Math.max(0, parseFloat((((revisedCost - originalCost) / originalCost) * 100).toFixed(1)))
      : 0;

    const expPct = latest.expenditure_pct_of_revised_cost ?? (
      revisedCost > 0 ? parseFloat(((latest.cumulative_expenditure_cr / revisedCost) * 100).toFixed(2)) : 0
    );
    const physPct = latest.physical_progress_pct ?? 0;
    const spendProgressGap = parseFloat((expPct - physPct).toFixed(2));

    const recommendedAction = buildRecommendedAction(
      riskAssessment,
      urgency,
      recentDeterioration,
      consecutiveStagnantMonths,
      costOverrunPercent,
      spendProgressGap
    );

    const priorityReason = buildPriorityReason(
      priorityTier,
      priorityScore,
      riskScore,
      urgency,
      recentDeterioration,
      evidenceConfidence,
      consecutiveStagnantMonths,
      topRiskDrivers
    );

    return {
      projectId: latest.project_id,
      projectName: latest.project_name,
      code: latest.legacy_ocms_code || (latest.pmgid ? `PMG-${latest.pmgid}` : `PAIMANA-${latest.project_id}`),
      sector: '', // Filled in by repository if derived sector available
      state: latest.state,
      implementingAgency: latest.agency,
      priorityScore,
      priorityTier,
      riskScore,
      riskTier,
      evidenceConfidence,
      urgency,
      recentDeterioration,
      primaryRiskDriver,
      topRiskDrivers,
      priorityReason,
      recommendedAction,
      evidence: {
        physicalProgressPercent: physPct,
        cumulativeExpenditureCr: latest.cumulative_expenditure_cr ?? 0,
        revisedCostCr: revisedCost,
        originalCostCr: originalCost,
        costOverrunPercent,
        expenditurePctOfRevisedCost: expPct,
        timeOverrunMonths: 0, // Filled in by repository
        targetCompletionDate: latest.revised_target_completion_mm_yyyy || latest.original_target_completion_mm_yyyy || 'Not specified',
        originalCompletionDate: latest.original_target_completion_mm_yyyy || 'Not specified',
        consecutiveStagnantMonths,
        recentProgressDelta: recentDelta,
        observationCount: obs.length
      },
      assessedAt: new Date().toISOString()
    };
  }
}

File: backend/services/priority_engine.py
from datetime import datetime, timezone
from typing import List, Optional, Tuple
from backend.models.project import PaimanaObservation
from backend.models.risk import RiskAssessment, PriorityAssessment, PriorityEvidence
from backend.models.common import PriorityTier
from backend.services.risk_engine import (
    PRISMRiskEngine, js_round, clamp,
    parse_completion_date_to_fractional_year,
    report_month_to_fractional_year
)

def calc_schedule_urgency(latest: PaimanaObservation) -> int:
    phys_pct = latest.physical_progress_pct if latest.physical_progress_pct is not None else 0.0
    if phys_pct >= 100.0:
        return 0

    target_date_str = latest.revised_target_completion_mm_yyyy or latest.original_target_completion_mm_yyyy
    target_year = parse_completion_date_to_fractional_year(target_date_str)
    if target_year is None:
        return 40  # Neutral baseline if no target completion date is provided

    report_year = report_month_to_fractional_year(latest.report_month or '2026-07')
    remaining_years = target_year - report_year

    if remaining_years <= 0:
        urgency = 85.0 + (100.0 - phys_pct) * 0.15
    elif remaining_years <= 0.5:
        time_factor = (1.0 - remaining_years / 0.5) * 15.0
        progress_factor = ((100.0 - phys_pct) / 100.0) * 10.0
        urgency = 70.0 + time_factor + progress_factor
    elif remaining_years <= 1.0:
        urgency = 50.0 + ((1.0 - remaining_years) / 0.5) * 20.0
    elif remaining_years <= 2.0:
        urgency = 30.0 + ((2.0 - remaining_years) / 1.0) * 20.0
    else:
        urgency = 30.0 - (remaining_years - 2.0) * 5.0

    return clamp(js_round(urgency), 0, 100)

def calc_recent_deterioration(obs: List[PaimanaObservation]) -> Tuple[int, int, Optional[float]]:
    if not obs or len(obs) < 2:
        return 30, 0, None

    deltas = [
        (obs[i].physical_progress_pct or 0.0) - (obs[i - 1].physical_progress_pct or 0.0)
        for i in range(1, len(obs))
    ]
    recent_delta = deltas[-1]

    consecutive_stagnant_months = 0
    for i in range(len(deltas) - 1, -1, -1):
        if deltas[i] <= 0.1:
            consecutive_stagnant_months += 1
        else:
            break

    if recent_delta < 0:
        score = 100.0
    elif recent_delta <= 0.1:
        if consecutive_stagnant_months >= 3:
            score = 95.0
        elif consecutive_stagnant_months == 2:
            score = 80.0
        else:
            score = 65.0
    elif recent_delta < 0.5:
        score = 55.0 if consecutive_stagnant_months > 0 else 40.0
    elif recent_delta < 1.0:
        score = 25.0
    elif recent_delta < 2.0:
        score = 10.0
    else:
        score = 0.0

    if len(obs) >= 3:
        early_deltas = deltas[:len(deltas) // 2]
        early_avg = sum(early_deltas) / (len(early_deltas) or 1)
        decel = early_avg - recent_delta
        if decel > 2.0:
            score += 15.0

    return (
        clamp(js_round(score), 0, 100),
        consecutive_stagnant_months,
        round(recent_delta, 2)
    )

def priority_tier_from_score(score: int) -> PriorityTier:
    if score >= 70:
        return 'P1'
    if score >= 50:
        return 'P2'
    return 'P3'

def build_recommended_action(
    assessment: RiskAssessment,
    urgency: int,
    recent_deterioration: int,
    consecutive_stagnant_months: int,
    cost_overrun_percent: float,
    spend_progress_gap: float
) -> str:
    actions: List[str] = []

    def get_indicator_score(id_name: str) -> int:
        for ind in assessment.indicators:
            if ind.id == id_name or ind.id.startswith(id_name):
                return ind.normalisedScore
        return 0

    stagnation_score = get_indicator_score('stagnation')
    cost_score = get_indicator_score('cost')
    schedule_score = get_indicator_score('schedule')
    divergence_score = get_indicator_score('divergence')
    velocity_score = get_indicator_score('velocity')

    # Rule 1: Prolonged Stagnation
    if consecutive_stagnant_months >= 2 or stagnation_score >= 60:
        actions.append('Escalate physical-progress review')

    # Rule 2: High Schedule Pressure
    if urgency >= 75 or schedule_score >= 60:
        actions.append('Review completion recovery plan')

    # Rule 3: Expenditure substantially ahead of physical progress (Divergence)
    if divergence_score >= 60 or spend_progress_gap >= 15:
        actions.append('Review expenditure–physical progress divergence')

    # Rule 4: Severe Cost Escalation
    if cost_score >= 60 or cost_overrun_percent >= 20:
        actions.append('Initiate cost/revision review')

    # Rule 5: Severe Velocity Deceleration
    if recent_deterioration >= 70 or velocity_score >= 70:
        if 'Escalate physical-progress review' not in actions:
            actions.append('Conduct contractor mobilization & site execution audit')

    if not actions:
        return 'Maintain routine milestone & expenditure monitoring'

    return '; '.join(actions)

def build_priority_reason(
    tier: PriorityTier,
    priority_score: int,
    risk_score: int,
    urgency: int,
    deterioration: int,
    confidence: float,
    consecutive_stagnant_months: int,
    top_drivers: List[str],
    cost_overrun_percent: float = 0.0,
    spend_progress_gap: float = 0.0,
    indicators: Optional[List[dict]] = None
) -> str:
    def get_ind_score(id_name: str) -> int:
        if not indicators:
            return 0
        for ind in indicators:
            ind_id = ind.get('id', '') if isinstance(ind, dict) else ind.id
            if ind_id == id_name or ind_id.startswith(id_name):
                return ind.get('normalisedScore', 0) if isinstance(ind, dict) else ind.normalisedScore
        return 0

    stagnation_score = get_ind_score('stagnation')
    velocity_score = get_ind_score('velocity')
    schedule_score = get_ind_score('schedule')
    cost_score = get_ind_score('cost')
    divergence_score = get_ind_score('divergence')

    if tier == 'P1':
        if consecutive_stagnant_months >= 2 or stagnation_score >= 75:
            if urgency >= 70 or schedule_score >= 70:
                return 'Progress has remained nearly stagnant across recent observations, compounded by schedule pressure.'
            return 'Progress has remained nearly stagnant across recent observations.'

        if (cost_overrun_percent >= 20 or cost_score >= 60) and (spend_progress_gap >= 15 or divergence_score >= 60):
            return 'Cost escalation and physical-financial divergence increase intervention urgency.'

        if risk_score >= 60 and (urgency >= 70 or schedule_score >= 65):
            return 'High risk combined with schedule pressure.'

        if deterioration >= 70 or velocity_score >= 75:
            return 'Recent progress velocity deterioration warrants immediate operational intervention.'

        if cost_overrun_percent >= 30 or cost_score >= 75:
            return 'Significant cost escalation over original outlay warrants urgent financial review.'

        if urgency >= 80:
            return 'Approaching or past target deadline with substantial physical work remaining.'

        primary = top_drivers[0] if top_drivers else 'Operational risk'
        return f"Critical {primary.lower()} combined with deadline urgency requires intervention."

    if tier == 'P2':
        if consecutive_stagnant_months >= 2 or stagnation_score >= 60:
            return 'Progress has remained nearly stagnant across recent observations.'
        if (cost_overrun_percent >= 15 or cost_score >= 50) and (spend_progress_gap >= 10 or divergence_score >= 50):
            return 'Cost escalation and physical-financial divergence increase intervention urgency.'
        if risk_score >= 55 and urgency >= 60:
            return 'High risk combined with schedule pressure.'
        if deterioration >= 50 or velocity_score >= 60:
            return 'Sluggish monthly progress velocity requires contractor mobilization oversight.'
        if urgency >= 65 or schedule_score >= 60:
            return 'Schedule slippage against target completion requires active timeline oversight.'
        primary = top_drivers[0] if top_drivers else 'Execution'
        return f"Moderate risk with notable {primary.lower()} monitoring requirements."

    return 'Routine monitoring: progress and expenditure remain within manageable variance thresholds.'

class PRISMPriorityEngine:
    """
    Authoritative PRISM Priority Engine (Phase 3A).
    Deterministically transforms risk assessments and urgency into the Intervention Priority Queue.
    """
    WEIGHTS = {
        'risk': 0.40,
        'urgency': 0.25,
        'deterioration': 0.20,
        'confidence': 0.15
    }

    @classmethod
    def assess(
        cls,
        observations: List[PaimanaObservation],
        existing_assessment: Optional[RiskAssessment] = None
    ) -> PriorityAssessment:
        if not observations:
            raise ValueError("PRISMPriorityEngine.assess() requires at least one observation")

        obs = sorted(observations, key=lambda o: o.report_month)
        latest = obs[-1]

        risk_assessment = existing_assessment or PRISMRiskEngine.assess(obs)
        risk_score = risk_assessment.riskScore
        risk_tier = risk_assessment.riskTier
        evidence_confidence = risk_assessment.evidenceConfidence

        urgency = calc_schedule_urgency(latest)
        recent_deterioration, consecutive_stagnant_months, recent_delta = calc_recent_deterioration(obs)
        confidence_score = clamp(evidence_confidence * 100.0, 0.0, 100.0)

        raw_priority_score = (
            risk_score * cls.WEIGHTS['risk'] +
            urgency * cls.WEIGHTS['urgency'] +
            recent_deterioration * cls.WEIGHTS['deterioration'] +
            confidence_score * cls.WEIGHTS['confidence']
        )

        priority_score = clamp(js_round(raw_priority_score), 0, 100)
        priority_tier = priority_tier_from_score(priority_score)

        sorted_indicators = sorted(
            risk_assessment.indicators,
            key=lambda ind: ind.weightedContribution,
            reverse=True
        )
        primary_risk_driver = sorted_indicators[0].label if sorted_indicators else 'General Execution'
        top_risk_drivers = [
            ind.label for ind in sorted_indicators if ind.normalisedScore >= 25
        ][:3]
        if not top_risk_drivers:
            top_risk_drivers.append(primary_risk_driver)

        original_cost = latest.original_cost_cr or 0.0
        revised_cost = latest.revised_cost_cr or original_cost
        cost_overrun_percent = (
            max(0.0, round(((revised_cost - original_cost) / original_cost) * 100.0, 1))
            if original_cost > 0 else 0.0
        )

        exp_pct = latest.expenditure_pct_of_revised_cost
        if exp_pct is None:
            exp_pct = round(((latest.cumulative_expenditure_cr or 0.0) / revised_cost) * 100.0, 2) if revised_cost > 0 else 0.0

        phys_pct = latest.physical_progress_pct or 0.0
        spend_progress_gap = round(exp_pct - phys_pct, 2)

        recommended_action = build_recommended_action(
            risk_assessment,
            urgency,
            recent_deterioration,
            consecutive_stagnant_months,
            cost_overrun_percent,
            spend_progress_gap
        )

        priority_reason = build_priority_reason(
            priority_tier,
            priority_score,
            risk_score,
            urgency,
            recent_deterioration,
            evidence_confidence,
            consecutive_stagnant_months,
            top_risk_drivers,
            cost_overrun_percent,
            spend_progress_gap,
            [ind.model_dump() for ind in risk_assessment.indicators]
        )

        return PriorityAssessment(
            projectId=latest.project_id,
            projectName=latest.project_name,
            code=latest.legacy_ocms_code or (f"PMG-{latest.pmgid}" if latest.pmgid else f"PAIMANA-{latest.project_id}"),
            sector="",
            state=latest.state,
            implementingAgency=latest.agency,
            priorityScore=priority_score,
            priorityTier=priority_tier,
            riskScore=risk_score,
            riskTier=risk_tier,
            evidenceConfidence=evidence_confidence,
            urgency=urgency,
            recentDeterioration=recent_deterioration,
            primaryRiskDriver=primary_risk_driver,
            topRiskDrivers=top_risk_drivers,
            priorityReason=priority_reason,
            recommendedAction=recommended_action,
            evidence=PriorityEvidence(
                physicalProgressPercent=phys_pct,
                cumulativeExpenditureCr=latest.cumulative_expenditure_cr or 0.0,
                revisedCostCr=revised_cost,
                originalCostCr=original_cost,
                costOverrunPercent=cost_overrun_percent,
                expenditurePctOfRevisedCost=exp_pct,
                timeOverrunMonths=0,
                targetCompletionDate=latest.revised_target_completion_mm_yyyy or latest.original_target_completion_mm_yyyy or 'Not specified',
                originalCompletionDate=latest.original_target_completion_mm_yyyy or 'Not specified',
                consecutiveStagnantMonths=consecutive_stagnant_months,
                recentProgressDelta=recent_delta,
                observationCount=len(obs)
            ),
            assessedAt=datetime.now(timezone.utc).isoformat()
        )
